File: metadata.py
import re
import json

WIKILINK_RE = re.compile(r'\[\[([^\]|]+)(?:\|([^\]]+))?\]\]')
HEADER_RE = re.compile(r'^(#{1,6})\s+(.*)', re.MULTILINE)
TASK_RE = re.compile(r'^- \[( |x|X)\] (.+)', re.MULTILINE)
TAG_RE = re.compile(r'#(\w+)')
LINK_RE = re.compile(r'\[([^\]]+)\]\(([^)]+)\)')
CODE_BLOCK_RE = re.compile(r'```(\w+)?\n(.*?)```', re.DOTALL)
IMAGE_RE = re.compile(r'!\[(.*?)\]\((.*?)\)')

def parse_markdown_file(filepath):
    with open(filepath, encoding='utf-8') as f:
        text = f.read()

    headers = []
    for m in HEADER_RE.finditer(text):
        level = len(m.group(1))
        header_text = m.group(2).strip()
        headers.append({'level': level, 'text': header_text})

    tasks = []
    for m in TASK_RE.finditer(text):
        checked = m.group(1).lower() == 'x'
        task_text = m.group(2).strip()
        tasks.append({'text': task_text, 'checked': checked})

    wikilinks = []
    for m in WIKILINK_RE.finditer(text):
        target = m.group(1)
        alias = m.group(2)
        wikilinks.append({'target': target, 'alias': alias})

    hyperlinks = []
    for m in LINK_RE.finditer(text):
        label = m.group(1)
        url = m.group(2)
        hyperlinks.append({'url': url, 'label': label})

    code_blocks = []
    for m in CODE_BLOCK_RE.finditer(text):
        lang = m.group(1) or 'plain'
        code_blocks.append({'language': lang})

    images = []
    for m in IMAGE_RE.finditer(text):
        alt_text = m.group(1)
        url = m.group(2)
        images.append({'alt_text': alt_text, 'url': url})

    tags_set = set(TAG_RE.findall(text))
    tags = [{'name': tag} for tag in sorted(tags_set)]

    metadata = {
        'tags': tags,
        'headers': headers,
        'tasks': tasks,
        'wikilinks': wikilinks,
        'hyperlinks': hyperlinks,
        'code_blocks': code_blocks,
        'images': images,
    }
    return metadata, text

File: cm_plugins/metadata.ts
// metadataStore.ts

import { Logger } from '../common/logger.ts';
const log = new Logger({ namespace: 'Metadata', minLevel: 'debug' });

export interface Header {
    level: number;
    text: string;
    line: number;
  }
export interface Tag{ //TODO: add page
    name: string;
    line: number;
    context: string;

}
  export interface Task {
    text: string;
    checked: boolean;
    line: number;
  }
  
  export interface CodeBlock {
    language: string | null;
    fromLine: number;
    toLine: number;
  }

  export interface Wikilink {
    target: string;       // Page reference, possibly with section: "Page#Section"
    alias?: string;       // Optional display text
    line: number;         // Line number where the link occurs
    context: string;      // Full line or surrounding content
  }
  
  export interface Hyperlink {
    url: string;
    label: string;
    line: number;
    context: string;
  }

  export interface Imagelink {
    url: string;
    altText: string;
    line: number;
    context: string;
  }
  
  export interface PageMetadata {
    lineCount: number;
    tags: Tag[];
    headers: Header[];
    tasks: Task[];
    codeBlocks: CodeBlock[];
    wikilinks: Wikilink[];
    hyperlinks: Hyperlink[];
    images: Imagelink[];
  }
  
  class MetadataStore {
    private tags: Tag[] = [];
    private headers: Header[] = [];
    private tasks: Task[] = [];
    private codeBlocks: CodeBlock[] = [];
    private wikilinks: Wikilink[] = [];
    private hyperlinks: Hyperlink[] = [];
    private imagelinks: Imagelink[] = [];
    private lineCount = 0;
  

    updateTags(newTags: Required<Tag>[]) {
        for (const newTag of newTags) {
          const newTagString = JSON.stringify(newTag);
      
          const exists = this.tags.some(tag => JSON.stringify(tag) === newTagString);
      
          if (!exists) {
            this.tags.push(newTag);
          }
        }
      
      }
      
    updateHeaders(newHeaders: Header[]) {
        for (const newHeader of newHeaders) {
            const newHeaderString = JSON.stringify(newHeader);
        
            const exists = this.headers.some(header => JSON.stringify(header) === newHeaderString);
        
            if (!exists) {
              this.headers.push(newHeader);
            }
          }
        
    }
    updateTasks(newTasks: Task[]) {
        for (const newTask of newTasks) {
            const newTaskString = JSON.stringify(newTask);
        
            const exists = this.tasks.some(tag => JSON.stringify(tag) === newTaskString);
        
            if (!exists) {
              this.tasks.push(newTask);
            }
          }
        
    }
    updateLineCount(lines: number) {
      this.lineCount = lines;
    }

    updateCodeBlocks(newBlocks: CodeBlock[]) {
      this.codeBlocks = newBlocks;
    }

    updateWikilinks(newLinks: Wikilink[]) {
      for (const link of newLinks) {
        const key = JSON.stringify(link);
        const exists = this.wikilinks.some(l => JSON.stringify(l) === key);
        if (!exists) {
          this.wikilinks.push(link);
        }
      }
    }
    
    updateHyperlinks(newLinks: Hyperlink[]) {
      for (const link of newLinks) {
        const linkString = JSON.stringify(link);
        const exists = this.hyperlinks.some(existing => JSON.stringify(existing) === linkString);
        if (!exists) {
          this.hyperlinks.push(link);
        }
      }
    }

    updateImages(newLinks: Imagelink[]) {
      for (const link of newLinks) {
        const linkString = JSON.stringify(link);
        const exists = this.imagelinks.some(existing => JSON.stringify(existing) === linkString);
        if (!exists) {
          this.imagelinks.push(link);
        }
      }
    }
  
    getMetadata(): PageMetadata {
      return {
        lineCount: this.lineCount,
        tags: this.tags,
        headers: this.headers,
        tasks: this.tasks,
        codeBlocks: this.codeBlocks,
        wikilinks: this.wikilinks,
        hyperlinks: this.hyperlinks,
        images: this.imagelinks
      };
    }
  }
  
  export const metadataStore = new MetadataStore();
  